executor/src/witgen/data_structures/finalizable_data.rs

```diff
@@ -0,0 +1,184 @@
+        let data = rows.map(Entry::InProgress).collect::<Vec<_>>();
+        Self { data, column_ids }
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn push(&mut self, row: Row<'a, T>) {
+        self.data.push(Entry::InProgress(row));
+    }
+
+    pub fn pop(&mut self) -> Option<Row<'a, T>> {
+        match self.data.pop() {
+            Some(Entry::InProgress(row)) => Some(row),
+            Some(Entry::Finalized(_, _)) => panic!("Row already finalized."),
+            None => None,
+        }
+    }
+
+    pub fn extend(&mut self, other: Self) {
+        self.data.extend(other.data);
+    }
+
+    pub fn remove(&mut self, i: usize) -> Row<'a, T> {
+        match self.data.remove(i) {
+            Entry::InProgress(row) => row,
+            Entry::Finalized(_, _) => panic!("Row {} already finalized.", i),
+        }
+    }
+
+    pub fn truncate(&mut self, len: usize) {
+        self.data.truncate(len);
+    }
+
+    pub fn get_mut(&mut self, i: usize) -> Option<&mut Row<'a, T>> {
+        match &mut self.data[i] {
+            Entry::InProgress(row) => Some(row),
+            Entry::Finalized(_, _) => panic!("Row {} already finalized.", i),
+        }
+    }
+
+    pub fn mutable_row_pair(&mut self, i: usize) -> (&mut Row<'a, T>, &mut Row<'a, T>) {
+        let (before, after) = self.data.split_at_mut(i + 1);
+        let current = before.last_mut().unwrap();
+        let next = after.first_mut().unwrap();
+        match (current, next) {
+            (Entry::InProgress(current), Entry::InProgress(next)) => (current, next),
+            _ => panic!("Row {} or {} (or both) already finalized.", i, i + 1),
+        }
+    }
+
+    pub fn finalize(&mut self, i: usize) {
+        if let Entry::InProgress(row) = &self.data[i] {
+            let (values, known_cells) = self
+                .column_ids
+                .iter()
+                .map(|c| (row[c].value.unwrap_or_default(), row[c].value.is_known()))
+                .unzip();
+            self.data[i] = Entry::Finalized(values, known_cells);
+        }
+    }
+
+    pub fn finalize_range(&mut self, range: impl Iterator<Item = usize>) {
+        for i in range {
+            self.finalize(i);
+        }
+    }
+
+    /// Takes all data out of the [FinalizableData] and returns it as a list of columns.
+    /// Columns are represented as a tuple of:
+    /// - A list of values
+    /// - A bit vector indicating which cells are known. Values of unknown cells should be ignored.
+    pub fn take_transposed(&mut self) -> impl Iterator<Item = (PolyID, (Vec<T>, BitVec))> {
+        let mut counter = 0;
+
+        log::info!(
+            "Transposing {} rows with {} columns...",
+            self.data.len(),
+            self.column_ids.len()
+        );
+        log::info!("Finalizing remaining rows...");
+        for i in 0..self.data.len() {
+            if matches!(self.data[i], Entry::InProgress(_)) {
+                self.finalize(i);
```

Comment:
You could change `finalize` to return a `bool` and call it unconditionally here.